keps/sig-scheduling/20190317-non-preempting-priorityclass.md

```diff
@@ -0,0 +1,116 @@
+---
+title: Add NonPreempting Option For PriorityClasses
+authors:
+  - "@vllry"
+owning-sig: sig-scheduling
+participating-sigs:
+  - sig-scheduling
+reviewers:
+  - "k82cn"
+  - "wgliang"
+approvers:
+  - "bsalamat"
+editor: Vallery Lancey
+creation-date: 2019-03-17
+last-updated: 2019-03-17
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Allow PriorityClasses To Be Non-Preempting
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+
+## Summary
+
+[PriorityClasses](https://kubernetes.io/docs/concepts/configuration/pod-priority-preemption/) are a GA feature as on 1.14,
+which impact the scheduling and eviction of pods.
+Pods will be scheduled according to descending priority.
+If a pod cannot be scheduled due to insufficient resources,
+lower-priority pods will be descheduled ("preempted") to make room.
+
+This proposal makes the preempting behavior optional for a PriorityClass,
+by adding a new field to PriorityClasses.
+If a PriorityClass does not have preemption enabled,
+a pod of that PriorityClass will not trigger preemption of other pods.
+
+## Motivation
+
+High-priority, non-preempting workloads are a common data science use case.
+Preempting batch workloads is a waste, as the work unit must be repeated.
+
+### Goals
+
+Add a boolean to PriorityClasses,
+to enable or disable preemption for pods of that PriorityClass.
+
+### Non-Goals
+
+## Proposal
+
+Add a Preempting field to PriorityClasses.
+This field will default to true,
+for backwards compatibility.
+
+```
+type PriorityClass struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+	Value int32
+	GlobalDefault bool
+	Description string
+	// New option
+	Preempting bool
+}
+```
+
+If Preempting is true for a pod,
+the scheduler will preempt lower priority pods to schedule this pod,
+as is current behavior.
+
+If Preempting is false,
+a pod of that class will not preempt other pods.
+
+Update our documentation to reflect this new feature.
+
+### Risks and Mitigations
+
+The new feature may malfuction,
+or preemption may be accidentally impaired.
+New tests (covering both nonpreepting workloads and mixed workloads),
+and the existing preempting PriorityClass tests should be used to prove stability.
+
+## Graduation Criteria
+
+* Users are reporting that this resolves their workload priority use-cases
+(if not, additional enhancements would be tightly linked to this one).
```

Comment:
Also adding use-cases section might be helpful.

Reply:
I mentioned the broad use case, but I'm not very familiar with with the workloads that require this.

Reply:
One of the more important use-case of this feature is for batch workloads. In batch processing, clusters often have much more work that they can fit simultaneously onto the nodes. So, there are many pods waiting in the queue to be scheduled. If a user has an important workload that needs to be scheduled before other pods in the queue, the user puts a higher priority on the workload to make it go to the head of the scheduling queue. At the same time, they don't want this workload to preempt running pods. Preemption is not desired for batch workloads, because a preempted pod often loses all it had done when it gets preempted and has to start from scratch next time it is scheduled. In order to achieve faster scheduling without preempting existing pods, the user can put a non-preempting high priority on their pods.